talk/morelanguage/exceptions.tex

```diff
@@ -4,185 +4,191 @@
+    \begin{cppcode*}{fontsize=\scriptsize,gobble=2}
+      try {
+        for (const &file : files) {
+          try {
+            process_file(file);
+          }
+          catch (const file_bad &e) {
+            ... // Skips one file
+          }
+        }
+      } catch (const db_bad &e) {
+        ... // This exception
+            // aborts everything
+      }
+    \end{cppcode*}
+    \columnbreak
+    \begin{cppcode*}{fontsize=\scriptsize,gobble=2}
+      int process_file(const fileobj file) {
+        if (fh = open_file(file)) {
+          throw
+            file_bad(file.error.reason());
+        }
+        while (!fh) {
+          line = read_line(fh);
+          add_line_to_db(line); // Can throw
+                                // db_bad
+        }
+        return 0;
+      }
+    \end{cppcode*}
+  \end{multicols}
+\end{frame}
+
+
+\begin{frame}[fragile]
+  \frametitlecpp[98]{Error Handling and Exceptions}
+  \begin{block}{}
     \begin{itemize}
-    \item throw specification is runtime only
+      \item exceptions have little cost if no exception is thrown
       \begin{itemize}
-      \item does not allow compiler optimizations
-      \item on the contrary forces extra checks
-      \item generally terminates your program if violated
+        \item a throw itself typically costs about as much as a function call (the stack needs to be unwound, but needs to be done for other error handling styles too)
+        \item so minimal if you are using exceptions for \textit{exceptional} events
+        \item thus their use is generally recommended
       \end{itemize}
-    \item throw specification clashes with templates
+      \item however, in a tight, performance-critical loop, where the error is handled close to the place it happens a more explicit error code or object might be better
       \begin{itemize}
-      \item one cannot ``template'' the throw clause
+        \item but profile if you really want to find out!
       \end{itemize}
-    \end{itemize}
-  \end{alertblock}
-\end{frame}
-
-\begin{frame}[fragile]
-  \frametitlecpp[11]{What remains}
-  \begin{block}{throw is dead}
-    \begin{itemize}
-    \item dynamic exception specification are deprecated
-    \item even throw() (no exceptions)
-    \end{itemize}
+   \end{itemize}
   \end{block}
-  \pause
-  \begin{exampleblock}{long live noexcept}
-    \begin{itemize}
-    \item noexcept a somehow equivalent to throw()
-    \item but is checked at compile time
-    \item so allows compiler optimizations
-    \end{itemize}
-  \end{exampleblock}
+  \begin{multicols}{2}
+    \begin{cppcode*}{fontsize=\tiny,gobble=2}
+      // try ... catch in hot loop
+      //   might impact performance
+      for (const &item: list) {
+        try {
+          check_item(item); // Can throw bad_item
+          ... // process item
+        }
+        catch (bad_item &e) {
+          ... // ignore this item, but keep going
+        }
+      }
+    \end{cppcode*}
+    \columnbreak
```

Comment:
I would like to see this in an `\begin{alertblock}{Don't}` (the red box) for discouraged things.